src/box/memtx_tree.cc

```diff
@@ -930,6 +965,120 @@ memtx_tree_index_get_internal(struct index *base, const char *key,
+	struct memtx_tree_index<USE_HINT> *index =
+		(struct memtx_tree_index<USE_HINT> *)it->index;
+	struct key_def *cmp_def = memtx_tree_cmp_def(&index->tree);
+	struct tree_iterator<USE_HINT> *tree_it =
+		get_tree_iterator<USE_HINT>(it);
+	struct tuple *tuple = tree_it->current.tuple;
+	if (tuple == NULL) {
+		*pos = NULL;
+		*size = 0;
+		return 0;
+	}
+	int mk_idx = MULTIKEY_NONE;
+	if (is_multikey)
+		mk_idx = tree_it->current.hint;
+	const char *key = tuple_extract_key(tuple, cmp_def, mk_idx, size);
+	if (key == NULL)
+		return -1;
+	*pos = key;
+	mp_decode_array(pos);
+	*size -= *pos - key;
+	return 0;
+}
+
+/**
+ * Implementation of @tree_iterator_position for functional indexes.
+ */
+static inline int
+tree_iterator_position_func_impl(struct iterator *it, const char **pos,
+				 uint32_t *size)
+{
+	/*
+	 * Сmp_def in the functional index is the functional key and the
+	 * primary key right after it. So to extract cmp_def in func index,
+	 * we need to pack an array with concatenated func key and primary key.
+	 */
+	struct tree_iterator<true> *tree_it = get_tree_iterator<true>(it);
+	if (tree_it->current_func_key == NULL) {
+		*pos = NULL;
+		*size = 0;
+		return 0;
+	}
+	assert(tree_it->current.tuple != NULL);
+	/* Extract func key. */
+	uint32_t func_key_size;
+	const char *func_key = tuple_data_range(tree_it->current_func_key,
+						&func_key_size);
+	uint32_t func_key_len = mp_decode_array(&func_key);
+	/* Extract primary key. */
+	struct space *space = space_by_id(it->index->def->space_id);
+	assert(space != NULL);
+	struct index *pk = space->index[0];
+	assert(pk != NULL);
+	struct key_def *pk_def = pk->def->key_def;
+	uint32_t pk_size;
+	const char *pk_key = tuple_extract_key(tree_it->current.tuple, pk_def,
+					       MULTIKEY_NONE, &pk_size);
+	uint32_t pk_key_len = mp_decode_array(&pk_key);
+	/* Calculate allocation size and allocate buffer. */
+	func_key_size -= mp_sizeof_array(func_key_len);
+	pk_size -= mp_sizeof_array(pk_key_len);
+	uint32_t alloc_size = func_key_size + pk_size;
+	char *data = (char *)region_alloc(&fiber()->gc, alloc_size);
+	if (data == NULL) {
+		diag_set(OutOfMemory, alloc_size, "region",
+			 "memtx_tree_func_index_iterator_position");
+		return -1;
+	}
+	*size = alloc_size;
+	*pos = data;
+	/* Pack an array with concatenated func key and primary key. */
+	memcpy(data, func_key, func_key_size);
+	data += func_key_size;
+	memcpy(data, pk_key, pk_size);
+	return 0;
+}
+
+/**
+ * Get position of iterator - extracted cmp_def of last fetched tuple.
+ * If iterator is exhausted, it returns position of last tuple, if there were
+ * no tuples fetched, empty position is returned.
+ */
+template <bool USE_HINT>
+static int
+tree_iterator_position(struct iterator *it, const char **pos, uint32_t *size)
+{
+	assert(it != NULL);
+	assert(pos != NULL);
+	assert(size != NULL);
+
+	struct key_def *key_def = it->index->def->key_def;
+	int rc = -1;
+	if (key_def->for_func_index) {
```

Comment:
`iterator_position` is a virtual method so we should set it to the right function pointer instead of doing branching inside.

Reply:
Moved this branching to `memtx_tree_create_iterator` method (since it is virtual method of iterator, we cannot get rid of branching at all).